docs/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/integrate-your-charm-with-postgresql.md

```diff
@@ -359,6 +359,166 @@ This should produce something similar to the output below (of course, with the n
 
 Congratulations, your relation with PostgreSQL is functional!
 
+## Add unit tests for your charm
+
+Now that our charm will extract the database authentication data and endpoints information from the relation data in function `fetch_postgres_relation_data`, we can add a test case to cover this scenario. In the test case, we create a relation and set it in the input state; on the relation-changed event, the database information should be extracted and set as environment variables for our `fastapi-service`:
+
+```python
+def test_relation_data():
+    ctx = testing.Context(FastAPIDemoCharm)
+    relation = testing.Relation(
+        endpoint="database",
+        interface="postgresql_client",
+        remote_app_name="postgresql-k8s",
+        remote_app_data={
+            "endpoints": "example.com:5432",
+            "username": "foo",
+            "password": "bar",
+        },
+    )
+    container = testing.Container(name="demo-server", can_connect=True)
+    state_in = testing.State(
+        containers={container},
+        relations={relation},
+        leader=True,
+    )
+
+    ctx.run(ctx.on.relation_changed(relation), state_in)
+
+    assert container.layers["fastapi_demo"].services["fastapi-service"].environment == {
+        "DEMO_SERVER_DB_HOST": "example.com",
+        "DEMO_SERVER_DB_PORT": "5432",
+        "DEMO_SERVER_DB_USER": "foo",
+        "DEMO_SERVER_DB_PASSWORD": "bar",
+    }
+```
+
+In this chapter, we have also defined a new method `_on_collect_status` that checks various things including if the required database relation is there or not. If not, we wait for the database relation and set it to the blocked status. We can also add a test case to cover this scenario by not setting a relation in the input state and assert the unit status is blocked:
+
+```python
+def test_no_database_blocked():
+    ctx = testing.Context(FastAPIDemoCharm)
+    container = testing.Container(name="demo-server", can_connect=True)
+    state_in = testing.State(
+        containers={container},
+        leader=True,
+    )
+
+    state_out = ctx.run(ctx.on.collect_unit_status(), state_in)
+
+    assert state_out.unit_status == ops.BlockedStatus("Waiting for database relation")
+```
+
+Run `tox -e unit` to make sure all test cases pass.
+
+## Add integration tests for your charm
+
+Now that our charm integrates with the PostgreSQL database, without a database integration, the app will be in the blocked status instead of active, because it requires a database relation. Let's tweak our existing integration test case `test_build_and_deploy` accordingly:
```

Comment:
It's probably easier on people if we say what we're changing here (it's just the `raise_on_blocked` I think?). There is the comment in the code, but that isn't as immediately obvious as saying something in the text.